File: apps/android/core/client/src/main/kotlin/social/waddle/android/client/ComposerMarkdown.kt
package social.waddle.android.client

import kotlinx.serialization.Serializable

/**
 * Send-time markdown: converts the literal markdown the user typed
 * into a clean body plus XEP-0394 markup spans, producing the same
 * wire shapes the web's WYSIWYG composer serializes
 * (`chat/src/lib/rich-message/serialize.ts`):
 *
 * - `**bold**`, `*italic*`, `~~strike~~`, `` `code` `` become inline
 *   spans with the marker characters removed from the body;
 * - ```` ``` ```` fences become `bcode` blocks — the fence lines are
 *   removed, the code text stays verbatim (the FFI span carries no
 *   language, so any fence info string is dropped);
 * - `>`-prefixed line groups become `bquote` blocks — the `>` markers
 *   STAY in the body (web wire parity; receivers strip them);
 * - lists are NOT markdown (web parity: XEP-0394 lists never hit the
 *   wire);
 * - all offsets are Unicode code points over the final body, and
 *   mention offsets recorded over the raw draft are rebased across
 *   every marker removal.
 *
 * Markers inside code (span or block) are literal. Marker styles do
 * not nest (`***x***` renders bold with literal inner asterisks); the
 * receiving renderer resolves one style per offset anyway.
 */

/** Serializable XEP-0394 span kind for queued offline sends. */
@Serializable
enum class MarkupRefType { BOLD, ITALIC, STRIKETHROUGH, CODE, CODE_BLOCK, BLOCKQUOTE }

/**
 * One outbound markup span: [begin] inclusive / [end] exclusive,
 * Unicode CODE POINTS over the trimmed send body BEFORE any reply
 * fallback prefix (`preparedSend` rebases onto the final wire body,
 * exactly like [MentionRef]).
 */
@Serializable
data class MarkupRef(val type: MarkupRefType, val begin: UInt, val end: UInt)

/** Result of [composeMarkdown]: clean body + spans + rebased mentions. */
data class ComposedMarkdown(
    val body: String,
    val markup: List<MarkupRef>,
    val mentions: List<MentionRef>,
)

/** Convert markdown in the trimmed [body]; [mentions] are rebased. */
fun composeMarkdown(body: String, mentions: List<MentionRef>): ComposedMarkdown {
    val fences = extractFences(body)
    val inline = applyInlineMarkers(fences.body, fences.codeBlocks)
    val codeBlocks = fences.codeBlocks.map { CpRange(inline.mapOffset(it.start), inline.mapOffset(it.end)) }
    val markup = codeBlocks.map { it.toMarkupRef(MarkupRefType.CODE_BLOCK) } +
        inline.spans +
        blockquoteSpans(inline.body, codeBlocks)
    val rebased = mentions.mapNotNull { mention ->
        val begin = inline.mapOffset(fences.mapOffset(mention.begin.toLong()))
        val end = inline.mapOffset(fences.mapOffset(mention.end.toLong()))
        if (end <= begin) null else mention.copy(begin = begin.toUInt(), end = end.toUInt())
    }
    return ComposedMarkdown(
        body = inline.body,
        markup = markup.sortedWith(compareBy({ it.begin }, { it.end })),
        mentions = rebased,
    )
}

// ── Shared primitives ────────────────────────────────────────────────

/** Code-point range, [start] inclusive, [end] exclusive. */
private data class CpRange(val start: Long, val end: Long) {
    fun overlaps(other: CpRange): Boolean = start < other.end && other.start < end
    fun contains(offset: Long): Boolean = offset in start until end
    fun toMarkupRef(type: MarkupRefType) =
        MarkupRef(type = type, begin = start.toUInt(), end = end.toUInt())
}

/** Removal of [length] code points at pre-removal offset [start]. */
private data class Removal(val start: Long, val length: Long)

/** Map a pre-removal offset to its post-removal position. */
private fun List<Removal>.mapOffset(offset: Long): Long {
    var shifted = offset
    for (removal in this) {
        if (offset <= removal.start) break
        shifted -= minOf(removal.length, offset - removal.start)
    }
    return maxOf(0, shifted)
}

// ── Pass A: fenced code blocks ───────────────────────────────────────

private class FencePass(
    val body: String,
    /** Code-block content ranges over [body]. */
    val codeBlocks: List<CpRange>,
    private val removals: List<Removal>,
) {
    fun mapOffset(offset: Long): Long = removals.mapOffset(offset)
}

private fun extractFences(body: String): FencePass {
    val lines = body.split("\n")
    val keep = BooleanArray(lines.size) { true }
    val contentGroups = mutableListOf<IntRange>()
    var opener = -1
    lines.forEachIndexed { index, line ->
        when {
            opener < 0 && line.startsWith("```") -> opener = index
            opener >= 0 && line.trim() == "```" -> {
                keep[opener] = false
                keep[index] = false
                if (index > opener + 1) contentGroups += (opener + 1) until index
                opener = -1
            }
        }
    }
    return assembleFencePass(lines, keep, contentGroups)
}

private fun assembleFencePass(
    lines: List<String>,
    keep: BooleanArray,
    contentGroups: List<IntRange>,
): FencePass {
    val lineCp = lines.map { it.codePointCount(0, it.length).toLong() }
    val oldStarts = LongArray(lines.size)
    for (index in 1 until lines.size) {
        oldStarts[index] = oldStarts[index - 1] + lineCp[index - 1] + 1
    }
    val removals = mutableListOf<Removal>()
    val kept = StringBuilder()
    val newStarts = LongArray(lines.size) { -1L }
    var newOffset = 0L
    lines.forEachIndexed { index, line ->
        if (keep[index]) {
            if (kept.isNotEmpty()) {
                kept.append('\n')
                newOffset += 1
            }
            newStarts[index] = newOffset
            kept.append(line)
            newOffset += lineCp[index]
        } else {
            // The dropped fence line plus the newline that joined it.
            removals += Removal(start = oldStarts[index], length = lineCp[index] + 1)
        }
    }
    val codeBlocks = contentGroups.map { group ->
        CpRange(newStarts[group.first], newStarts[group.last] + lineCp[group.last])
    }
    return FencePass(body = kept.toString(), codeBlocks = codeBlocks, removals = removals)
}

// ── Pass B: inline markers, single left-to-right scan ───────────────

private class InlinePass(
    val body: String,
    val spans: List<MarkupRef>,
    private val removals: List<Removal>,
) {
    fun mapOffset(offset: Long): Long = removals.mapOffset(offset)
}

private class InlineRule(pattern: String, val type: MarkupRefType, val markerLength: Int) {
    val regex = Regex(pattern)
}

/** Priority order at equal positions: code wins, then strike, bold, italic. */
private val INLINE_RULES = listOf(
    InlineRule("`([^`\n]+)`", MarkupRefType.CODE, markerLength = 1),
    InlineRule("~~([^\n]+?)~~", MarkupRefType.STRIKETHROUGH, markerLength = 2),
    // The optional tail group is lazy (`??`) so `**a** mid **b**`
    // yields two spans instead of one greedy span across the middle.
    InlineRule("""\*\*(\S(?:[^\n]*?\S)??)\*\*""", MarkupRefType.BOLD, markerLength = 2),
    InlineRule("""\*(\S(?:[^*\n]*?\S)??)\*""", MarkupRefType.ITALIC, markerLength = 1),
)

private data class InlineMatch(val rule: InlineRule, val match: MatchResult)

private fun applyInlineMarkers(body: String, protectedRanges: List<CpRange>): InlinePass {
    val result = StringBuilder()
    val spans = mutableListOf<MarkupRef>()
    val removals = mutableListOf<Removal>()
    var searchFrom = 0
    var removedSoFar = 0L
    while (searchFrom < body.length) {
        val next = earliestInlineMatch(body, searchFrom, protectedRanges) ?: break
        val match = next.match
        val content = match.groupValues[1]
        val markerCp = next.rule.markerLength.toLong()
        val startCp = body.codePointCount(0, match.range.first).toLong()
        val endCp = body.codePointCount(0, match.range.last + 1).toLong()
        result.append(body, searchFrom, match.range.first)
        result.append(content)
        removals += Removal(start = startCp, length = markerCp)
        removals += Removal(start = endCp - markerCp, length = markerCp)
        val spanStart = startCp - removedSoFar
        val contentCp = content.codePointCount(0, content.length).toLong()
        spans += CpRange(spanStart, spanStart + contentCp).toMarkupRef(next.rule.type)
        removedSoFar += markerCp * 2
        searchFrom = match.range.last + 1
    }
    result.append(body, searchFrom, body.length)
    return InlinePass(body = result.toString(), spans = spans, removals = removals)
}

private fun earliestInlineMatch(
    body: String,
    searchFrom: Int,
    protectedRanges: List<CpRange>,
): InlineMatch? {
    var from = searchFrom
    while (from < body.length) {
        val candidates = INLINE_RULES.mapNotNull { rule ->
            rule.regex.find(body, from)?.let { InlineMatch(rule, it) }
        }
        val best = candidates.minByOrNull { it.match.range.first } ?: return null
        val startCp = body.codePointCount(0, best.match.range.first).toLong()
        val endCp = body.codePointCount(0, best.match.range.last + 1).toLong()
        if (protectedRanges.none { it.overlaps(CpRange(startCp, endCp)) }) return best
        // Protected (inside a code block): retry past this match start.
        from = best.match.range.first + 1
    }
    return null
}

// ── Pass C: blockquote line groups ───────────────────────────────────

private fun blockquoteSpans(body: String, codeBlocks: List<CpRange>): List<MarkupRef> {
    val spans = mutableListOf<MarkupRef>()
    var offset = 0L
    var groupStart = -1L
    var groupEnd = -1L
    fun flush() {
        if (groupStart >= 0) {
            spans += CpRange(groupStart, groupEnd).toMarkupRef(MarkupRefType.BLOCKQUOTE)
            groupStart = -1
        }
    }
    for (line in body.split("\n")) {
        val lineCp = line.codePointCount(0, line.length).toLong()
        val insideCode = codeBlocks.any { it.contains(offset) }
        if (line.startsWith(">") && !insideCode) {
            if (groupStart < 0) groupStart = offset
            groupEnd = offset + lineCp
        } else {
            flush()
        }
        offset += lineCp + 1
    }
    flush()
    return spans
}

File: apps/apple/WaddleKit/Sources/WaddleKit/Compose/ComposerMarkdown.swift
import Foundation

// Send-time markdown: converts the literal markdown the user typed into a
// clean body plus XEP-0394 markup spans, producing the same wire shapes as
// Android's `composeMarkdown` and the web's WYSIWYG composer
// (`chat/src/lib/rich-message/serialize.ts`):
//
// - `**bold**`, `*italic*`, `~~strike~~`, `` `code` `` become inline spans
//   with the marker characters removed from the body;
// - ```` ``` ```` fences become code blocks: the fence lines are removed,
//   the code text stays verbatim (the span carries no language, so any
//   fence info string is dropped);
// - `>`-prefixed line groups become blockquotes; the `>` markers STAY in
//   the body (web wire parity; receivers strip them);
// - lists are NOT markdown (XEP-0394 lists never hit the wire);
// - all offsets are Unicode scalars (code points) over the final body, and
//   mention offsets recorded over the raw draft are rebased across every
//   marker removal.
//
// Markers inside code (span or block) are literal. Marker styles do not
// nest (`***x***` renders bold with literal inner asterisks); the receiving
// renderer resolves one style per offset anyway.

/// Who a mention addresses (XEP-0372 URI shapes shared with web/Android).
public enum MentionTarget: Hashable, Sendable {
    case user(BareJID)
    case everyone
    case here

    /// `xmpp:<barejid>`, or the literal broadcast URIs `xmpp:@everyone` / `xmpp:@here`.
    public var uri: String {
        switch self {
        case let .user(jid): return "xmpp:\(jid)"
        case .everyone: return "xmpp:@everyone"
        case .here: return "xmpp:@here"
        }
    }
}

/// A mention the composer inserted: its target and its range over the raw
/// draft, in Unicode scalars (code points), end exclusive.
public struct MentionDraft: Hashable, Sendable {
    public let target: MentionTarget
    public var range: Range<Int>

    public init(target: MentionTarget, range: Range<Int>) {
        self.target = target
        self.range = range
    }
}

public struct ComposedMarkdown: Hashable, Sendable {
    public let body: String
    /// Sorted by (start, end). Kinds used: .bold, .italic, .strikethrough,
    /// .code, .codeBlock, .blockquote (never .link).
    public let spans: [MarkupSpan]
    /// Mentions rebased onto `body`; mentions that collapse to empty are dropped.
    public let mentions: [MentionDraft]
}

public enum ComposerMarkdown {
    /// Converts markdown in the trimmed `body`; `mentions` are rebased.
    public static func compose(_ body: String, mentions: [MentionDraft]) -> ComposedMarkdown {
        let fences = extractFences(Array(body.unicodeScalars))
        let inline = applyInlineMarkers(fences.body, protected: fences.codeBlocks)
        let codeBlocks = fences.codeBlocks.map {
            ScalarRange(start: inline.removals.mapOffset($0.start), end: inline.removals.mapOffset($0.end))
        }
        let spans = codeBlocks.map { $0.span(.codeBlock) }
            + inline.spans
            + blockquoteSpans(inline.body, codeBlocks: codeBlocks)
        return ComposedMarkdown(
            body: String(String.UnicodeScalarView(inline.body)),
            spans: sortedStably(spans),
            mentions: mentions.compactMap { rebase($0, fences: fences.removals, inline: inline.removals) }
        )
    }
}

// MARK: - Shared primitives

private typealias Scalars = [Unicode.Scalar]

/// Scalar range, `start` inclusive, `end` exclusive.
private struct ScalarRange {
    let start: Int
    let end: Int

    func overlaps(_ other: ScalarRange) -> Bool { start < other.end && other.start < end }
    func contains(_ offset: Int) -> Bool { start <= offset && offset < end }
    func span(_ kind: MarkupSpan.Kind) -> MarkupSpan { MarkupSpan(kind: kind, start: start, end: end) }
}

/// Removal of `length` scalars at pre-removal offset `start`.
private struct Removal {
    let start: Int
    let length: Int
}

extension [Removal] {
    /// Maps a pre-removal offset to its post-removal position. Removals are
    /// ascending by `start`.
    fileprivate func mapOffset(_ offset: Int) -> Int {
        var shifted = offset
        for removal in self {
            if offset <= removal.start { break }
            shifted -= Swift.min(removal.length, offset - removal.start)
        }
        return Swift.max(0, shifted)
    }
}

private func rebase(_ mention: MentionDraft, fences: [Removal], inline: [Removal]) -> MentionDraft? {
    let start = inline.mapOffset(fences.mapOffset(mention.range.lowerBound))
    let end = inline.mapOffset(fences.mapOffset(mention.range.upperBound))
    guard end > start else { return nil }
    return MentionDraft(target: mention.target, range: start..<end)
}

/// Sorts by (start, end), keeping insertion order on ties like Kotlin's
/// stable `sortedWith`.
private func sortedStably(_ spans: [MarkupSpan]) -> [MarkupSpan] {
    spans.enumerated()
        .sorted { lhs, rhs in
            (lhs.element.start, lhs.element.end, lhs.offset) < (rhs.element.start, rhs.element.end, rhs.offset)
        }
        .map(\.element)
}

private func splitLines(_ scalars: Scalars) -> [Scalars] {
    scalars.split(separator: "\n", omittingEmptySubsequences: false).map(Array.init)
}

private func hasPrefix(_ scalars: Scalars, _ prefix: Scalars, at index: Int) -> Bool {
    guard index >= 0, index + prefix.count <= scalars.count else { return false }
    return scalars[index..<(index + prefix.count)].elementsEqual(prefix)
}

// MARK: - Pass A: fenced code blocks

private struct FencePass {
    let body: Scalars
    /// Code-block content ranges over `body`.
    let codeBlocks: [ScalarRange]
    let removals: [Removal]
}

private let fence: Scalars = Array("```".unicodeScalars)

private func extractFences(_ body: Scalars) -> FencePass {
    let lines = splitLines(body)
    var keep = [Bool](repeating: true, count: lines.count)
    var contentGroups: [ClosedRange<Int>] = []
    var opener: Int?
    for (index, line) in lines.enumerated() {
        if let open = opener {
            guard trimmedKotlinStyle(line) == fence else { continue }
            keep[open] = false
            keep[index] = false
            if index > open + 1 { contentGroups.append((open + 1)...(index - 1)) }
            opener = nil
        } else if hasPrefix(line, fence, at: 0) {
            opener = index
        }
    }
    return assembleFencePass(lines, keep: keep, contentGroups: contentGroups)
}

private func assembleFencePass(_ lines: [Scalars], keep: [Bool], contentGroups: [ClosedRange<Int>]) -> FencePass {
    var removals: [Removal] = []
    var kept: Scalars = []
    var newStarts = [Int](repeating: -1, count: lines.count)
    var oldStart = 0
    for (index, line) in lines.enumerated() {
        if keep[index] {
            // Kotlin joins on `kept.isNotEmpty()`, so leading empty kept
            // lines contribute no newline; mirrored for wire parity.
            if !kept.isEmpty { kept.append("\n") }
            newStarts[index] = kept.count
            kept.append(contentsOf: line)
        } else {
            // The dropped fence line plus the newline that joined it.
            removals.append(Removal(start: oldStart, length: line.count + 1))
        }
        oldStart += line.count + 1
    }
    let codeBlocks = contentGroups.map { group in
        ScalarRange(start: newStarts[group.lowerBound], end: newStarts[group.upperBound] + lines[group.upperBound].count)
    }
    return FencePass(body: kept, codeBlocks: codeBlocks, removals: removals)
}

/// Kotlin `CharSequence.trim()`: `Character.isWhitespace || isSpaceChar`.
private func trimmedKotlinStyle(_ line: Scalars) -> Scalars {
    guard let first = line.firstIndex(where: { !isKotlinWhitespace($0) }),
          let last = line.lastIndex(where: { !isKotlinWhitespace($0) })
    else { return [] }
    return Array(line[first...last])
}

private func isKotlinWhitespace(_ scalar: Unicode.Scalar) -> Bool {
    (0x09...0x0D).contains(scalar.value) || (0x1C...0x1F).contains(scalar.value) || isSeparator(scalar)
}

private func isSeparator(_ scalar: Unicode.Scalar) -> Bool {
    switch scalar.properties.generalCategory {
    case .spaceSeparator, .lineSeparator, .paragraphSeparator: return true
    default: return false
    }
}

// MARK: - Pass B: inline markers, single left-to-right scan

private struct InlinePass {
    let body: Scalars
    let spans: [MarkupSpan]
    let removals: [Removal]
}

/// One marker style. `flanked` content is `\S(?:[^excluded]*?\S)??` (the
/// bold/italic shape); otherwise it is `[^excluded]+?`. Both are matched
/// lazily so `**a** mid **b**` yields two spans instead of one greedy span
/// across the middle.
private struct InlineRule {
    let marker: Scalars
    let kind: MarkupSpan.Kind
    let excluded: Set<Unicode.Scalar>
    let flanked: Bool
}

/// Priority order at equal positions: code wins, then strike, bold, italic.
private let inlineRules: [InlineRule] = [
    InlineRule(marker: ["`"], kind: .code, excluded: ["`", "\n"], flanked: false),
    InlineRule(marker: ["~", "~"], kind: .strikethrough, excluded: ["\n"], flanked: false),
    InlineRule(marker: ["*", "*"], kind: .bold, excluded: ["\n"], flanked: true),
    InlineRule(marker: ["*"], kind: .italic, excluded: ["*", "\n"], flanked: true),
]

private struct InlineMatch {
    let rule: InlineRule
    /// Whole match including both markers.
    let range: ScalarRange
    var content: ScalarRange {
        ScalarRange(start: range.start + rule.marker.count, end: range.end - rule.marker.count)
    }
}

private func applyInlineMarkers(_ body: Scalars, protected: [ScalarRange]) -> InlinePass {
    var result: Scalars = []
    var spans: [MarkupSpan] = []
    var removals: [Removal] = []
    var searchFrom = 0
    while searchFrom < body.count, let match = earliestInlineMatch(body, from: searchFrom, protected: protected) {
        let marker = match.rule.marker.count
        let content = match.content
        result.append(contentsOf: body[searchFrom..<match.range.start])
        let spanStart = result.count
        result.append(contentsOf: body[content.start..<content.end])
        spans.append(ScalarRange(start: spanStart, end: result.count).span(match.rule.kind))
        removals.append(Removal(start: match.range.start, length: marker))
        removals.append(Removal(start: content.end, length: marker))
        searchFrom = match.range.end
    }
    result.append(contentsOf: body[Swift.min(searchFrom, body.count)...])
    return InlinePass(body: result, spans: spans, removals: removals)
}

/// Leftmost match, ties broken by rule priority. A match overlapping a code
/// block is skipped by retrying one scalar past its start.
private func earliestInlineMatch(_ body: Scalars, from: Int, protected: [ScalarRange]) -> InlineMatch? {
    var position = from
    while position < body.count {
        if let match = firstRuleMatch(body, at: position) {
            if !protected.contains(where: { $0.overlaps(match.range) }) { return match }
        }
        position += 1
    }
    return nil
}

private func firstRuleMatch(_ body: Scalars, at start: Int) -> InlineMatch? {
    for rule in inlineRules {
        if let end = matchEnd(rule, in: body, at: start) {
            return InlineMatch(rule: rule, range: ScalarRange(start: start, end: end))
        }
    }
    return nil
}

/// End (exclusive, after the closing marker) of `rule` matched at `start`.
private func matchEnd(_ rule: InlineRule, in body: Scalars, at start: Int) -> Int? {
    guard hasPrefix(body, rule.marker, at: start) else { return nil }
    let contentStart = start + rule.marker.count
    let contentEnd = rule.flanked
        ? flankedContentEnd(rule, in: body, from: contentStart)
        : plainContentEnd(rule, in: body, from: contentStart)
    return contentEnd.map { $0 + rule.marker.count }
}

/// Lazy `[^excluded]+?` followed by the closing marker.
private func plainContentEnd(_ rule: InlineRule, in body: Scalars, from contentStart: Int) -> Int? {
    var end = contentStart + 1
    while end <= body.count {
        if rule.excluded.contains(body[end - 1]) { return nil }
        if hasPrefix(body, rule.marker, at: end) { return end }
        end += 1
    }
    return nil
}

/// Lazy `\S(?:[^excluded]*?\S)??` followed by the closing marker: the
/// content starts and ends on non-space, and every scalar between is
/// outside `excluded`.
private func flankedContentEnd(_ rule: InlineRule, in body: Scalars, from contentStart: Int) -> Int? {
    guard contentStart < body.count, !isRegexSpace(body[contentStart]) else { return nil }
    var end = contentStart + 1
    while end <= body.count {
        if end >= contentStart + 3, rule.excluded.contains(body[end - 2]) { return nil }
        if !isRegexSpace(body[end - 1]), hasPrefix(body, rule.marker, at: end) { return end }
        end += 1
    }
    return nil
}

/// Regex `\s`. The Kotlin tests run on the JVM (ASCII `[ \t\n\x0B\f\r]`)
/// while Android devices use ICU (`[\t\n\f\r\p{Z}]`); the union keeps both
/// ASCII controls and Unicode separators from flanking a style.
private func isRegexSpace(_ scalar: Unicode.Scalar) -> Bool {
    scalar == " " || (0x09...0x0D).contains(scalar.value) || isSeparator(scalar)
}

// MARK: - Pass C: blockquote line groups

private func blockquoteSpans(_ body: Scalars, codeBlocks: [ScalarRange]) -> [MarkupSpan] {
    var spans: [MarkupSpan] = []
    var offset = 0
    var group: ScalarRange?
    for line in splitLines(body) {
        let insideCode = codeBlocks.contains { $0.contains(offset) }
        if line.first == ">", !insideCode {
            group = ScalarRange(start: group?.start ?? offset, end: offset + line.count)
        } else if let finished = group {
            spans.append(finished.span(.blockquote))
            group = nil
        }
        offset += line.count + 1
    }
    if let finished = group { spans.append(finished.span(.blockquote)) }
    return spans
}
